### ISSUE TITLE
fireqos' match tcp syn filter

### ISSUE THREAD

**Issue by ooptimum** [firehol/firehol#330]
"match tcp syn" statement translates to tc filter "u32 match u8 0x02 0x02 at 33", i.e. with no check of protocol at all. That is it matches any packet with 2nd bit of a byte at offset 33 set to 1 regardless of the protocol.
Correct rule should look something like this: "u32 match ip protocol 6 0xff match u8 0x02 0x02 at 33".

**Comment by thomric2** [comment]
I just came across this today as well.

A match tcp syn is just 
`tc filter add dev ${device} parent 1:12 protocol ip prio 130 u32 match u8 0x02 0x02 at 33 flowid 1:17`

While a match tcp ack is
`tc filter add dev ${device} parent 1:12 protocol ip prio 140 u32 match ip protocol 6 0xff match u8 0x05 0x0f at 0 match u16 0x0000 0xffc0 at 2 match u8 0x10 0xff at 33 flowid 1:17`

So the ack also compares the protocol as @ooptimum mentioned was missing from the syn but then also adds in 3 more compares:

1. IP packet header length is 20
2. IP packet size less than 64
3. The only TCP flag set is ACK

I'm not sure that #3 should be exclusive with the mask of 0xff instead of 0x10 when there is already the rule to limit packets to just size 64 and less.

And then I'm guessing that since packets with the SYN flag set don't have payloads there is no reason to add those checks like there are in the ACK rules.

Though you may also want the IP Packet Header length is 20 match for "match tcp syn" since it is relying on the 33rd byte which is after the IP header and if the header is > 20 then the match will not line up.

In summary the "match tcp syn" should probably generate 
`u32 match ip protocol 6 0xff match u8 0x05 0x0f at 0 match u8 0x02 0x02 at 33`
